### PR TITLE
Add Doc Blocks overview & API reference

### PR DESCRIPTION
**Pull request by kylegach**
## What I did

- Add Doc Blocks overview & API reference pages for each block
- Remove Doc Blocks guides
- Update `docs/api/argtypes` to include information that was in `docs/writing-docs/doc-block-argstable`
- Update TOC
    - Change `Write docs > Doc blocks` from a menu to a page
    - Add `API > @storybook/blocks menu`

<table>
<tr>
	<th>Before
	<th>After
<tr>
	<td>Screenshot of current TOC
	<td>Screenshot of proposed TOC
</table>

### Example page

Screenshot of ArgTypes doc block API reference

### WIP

- Add/update screenshots for all doc blocks

## How to test

1. In `frontpage`, check out the `docs-content-style-tweaks` branch ([see related PR](https://github.com/storybookjs/frontpage/pull/487))
    - The content won't look right otherwise
2. Extract the docs for this branch: `yarn extract-monorepo-docs doc-blocks-docs`
3. Start the site: `GATSBY_SKIP_ADDON_PAGES=true yarn start`
4. Navigate to the updated/new pages:
    - [Doc blocks overview](http://localhost:8000/docs/react/writing-docs/doc-blocks)
    - [ArgTypes API reference](http://localhost:8000/docs/react/api/argtypes/)
    - And so on...

### What to check for:

- Correctness
- Consistent, helpful structure
- Understandability of text content
- _**Not**_ design ([separate `frontpage` PR for that](https://github.com/storybookjs/frontpage/pull/487))

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [x] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: docs/api/doc-block-argtypes.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+`ArgTypes` is a React component which accepts props of type `ArgTypesProps`.
+
+<div class="aside">
+
+ℹ️ Like most blocks, the `ArgTypes` block can both be configured via props when using it directly in MDX, or with properties in `parameters.docs.argTypes`.
+
+</div>
+
+#### `ArgTypesProps`
```
Comment (JReinhold) [inline docs/api/doc-block-argtypes.md:41]: Personally I feel the pattern of documenting `ArgTypes` separately from `ArgTypesProps` is very redundant (in all cases, not just for `ArgTypes`. I don't really see how this adds any value/clarification, compared to this simplification:

```suggestion

`ArgTypes` is a React component which accepts the following props:

<div class="aside">

ℹ️ Like most blocks, the `ArgTypes` block can both be configured via props when using it directly in MDX, or with properties in `parameters.docs.argTypes`.

</div>

```

File: docs/api/doc-block-markdown.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+```md
+# A header
+
+{ this is valid in a plain Markdown file, but MDX2 will try to evaluate this as an expression }
+
+<This is also valid, but MDX2 thinks this is a JSX component />
+```
+<!-- prettier-ignore-end -->
+
+Furthermore MDX2 wraps all strings on newlines in p tags or similar, meaning that content would render different between a plain .md file and an .mdx file.
```
Comment (JReinhold) [inline docs/api/doc-block-markdown.md:74]:
```suggestion
Furthermore MDX2 wraps all strings on newlines in `p` tags or similar, meaning that content would render different between a plain .md file and an .mdx file.
```

File: docs/api/doc-block-markdown.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+// DO this, will work
+import ReadMe from './README.md?raw';
+
+import { Markdown } from '@storybook/blocks';
+
+# A header 
+
+<Markdown>{ReadMe}</Markdown />
+```
+<!-- prettier-ignore-end -->
```
Comment (JReinhold) [inline docs/api/doc-block-markdown.md:22]: This paragraph from the drafts are omitted here (on purpose or not?), but I think it's an important companion to the snippet.

> When importing markdown files it’s important to use the ?raw suffix on the import path to ensure the content is imported as-is, instead of being evaluated:

```suggestion
The `Markdown` block allows you to import and include plain markdown in your MDX files.

![Screenshot of Markdown block](TK)

When importing markdown files it’s important to use the `?raw` suffix on the import path to ensure the content is imported as-is, instead of being evaluated:

<!-- prettier-ignore-start -->
```md
// DON'T do this, will error
import ReadMe from './README.md';
// DO this, will work
import ReadMe from './README.md?raw';

import { Markdown } from '@storybook/blocks';

# A header 

<Markdown>{ReadMe}</Markdown />
```

```

File: docs/api/doc-block-story.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+Type: `StoryName`
+
+Defining and configuring stories in MDX is deprecated. See the [Migration guide](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#mdx-docs-files) for details.
+
+##### `tags` (deprecated)
+
+Type: `Tag[]`
+
+Defining and configuring stories in MDX is deprecated. See the [Migration guide](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#mdx-docs-files) for details.
```
Comment (JReinhold) [inline docs/api/doc-block-story.md:164]: I don't think we need to document this "deprecated" prop, given that tags is only being introduced in 7.0.

```suggestion
```

**Review by JReinhold — CHANGES_REQUESTED** [review]
Fantastic job @kylegach ! I have a few suggestions all over the place.

File: docs/api/argtypes.md [hunk 5]
```diff
@@ -57,37 +107,74 @@ This ArgTypes data structure, name, type, defaultValue, and description are stan
 
 </div>
 
-#### Manual specification
+### Grouping
 
-If you want more control over the args table or any other aspect of using argTypes, you can overwrite the generated argTypes for your component on a per-arg basis. For instance, with the above-inferred argTypes and the following default export:
+You can also extend the ArgsTable's customization by grouping related `argTypes` into categories or even subcategories. Based on the following component implementation:
```
[the remark this answers is not in the document]
Comment (kylegach) [reply docs/api/argtypes.md:112]: Good spot! This is lazy copy/pasting from existing 6.5 docs.

File: docs/api/doc-block-canvas.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+Specifies the initial state of the source panel.
+
+- **hidden**: the source panel is hidden by default
+- **shown**: the source panel is shown by default
+- **none**: the source panel is not available and the button to show it is not rendered
+
+##### `story`
+
+Type: `Pick<StoryProps, 'inline' | 'height' | 'autoplay'>`
```
[the remark this answers is not in the document]
Comment (kylegach) [reply docs/api/doc-block-canvas.md:163]: Yeah, I was trying to emulate what something like `react-docgen-typescript` would provide for us. But you're probably right that it's better to be simpler and more explicit.

I'll draw the line at including links in the `Type:` line, though, as that will very likely _never_ be possible with any tooling. I'll make sure the description links to the Story block's options.

File: docs/api/doc-block-colorpalette.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+`ColorItem` is a React component which accepts props of type `ColorItemProps`.
+
+#### `ColorItemProps`
+
+##### `colors` (required)
+
+Type: `string[] | { [key: string]: string }`
+
+Provides the list of colors to be displayed.
```
[the remark this answers is not in the document]
Comment (kylegach) [reply docs/api/doc-block-colorpalette.md:52]: Looking at the [source](https://github.com/storybookjs/storybook/blob/next/code/ui/blocks/src/components/ColorPalette.tsx), it literally just passes the strings to apply as `background: <value>`. So it supports any valid CSS color format. I don't mind making that more clear, though.

File: docs/api/doc-block-description.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+See [Migration guide](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#description-block-parametersnotes-and-parametersinfo).
+
+##### `type` (deprecated)
+
+Type: `DescriptionType`
+
+See [Migration guide](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#description-block-parametersnotes-and-parametersinfo).
+
+## Writing descriptions
```
[the remark this answers is not in the document]
Comment (kylegach) [reply docs/api/doc-block-description.md:63]: Hmm... I agree. But this section of text is largely about the various ways to provide a description and how they cascade. I think we should leave this here, as-is. Then, later, we can add a section that references this one to either an existing guide (like the upcoming AutoDocs page) or a new one (_Documentation best practices_, maybe?). My vote is the latter.

File: docs/api/doc-block-argtypes.md (hunk 1 above)
Comment (kylegach) [reply docs/api/doc-block-argtypes.md:41]: This is largely driven by pages like `/docs/api/doc-block-colorpalette`, which contains both `ColorPalette` and `ColorItem`, each with their own props.

However, I now see why your Notion examples used "Options" as a heading (to be inclusive of both props and parameters uses), and I'll be switching to that.

File: docs/api/doc-block-markdown.md (hunk 3 above)
Comment (kylegach) [reply docs/api/doc-block-markdown.md:22]: Whoops! Thanks for the catch. Agreed it's important.

File: docs/api/doc-block-useof.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+- `useOf('story')` returns the annotated primary story in attached mode, error in unattached mode
+- `useOf('meta')` returns the annotated meta in attached mode, error in unattached mode
+- `useOf('component')` returns the annotated component specified in the meta in attached mode, error in unattached mode
+
+### `validTypes`
+
+Type: `Array<'story' | 'meta' | 'component'>`
+
+Optionally specify an array of valid types that your block accepts. Ie. the Canvas block only accepts references to stories, not meta or component, so it has `useOf(of, ['story'])` . If anything else than the valid types are passed in, an error will be thrown.
```
Comment (jonniebigodes) [inline docs/api/doc-block-useof.md:85]: When able rephrase this to "Optionally specify an array of valid types that your block accepts. Anything else that the valid types will result in an error. For example, the `Canvas` doc only accepts references to stories, not meta or component definitions. With that, it uses a `useOf(of, ['story'])` hook implementation." for a bit more clarity.

**Review by jonniebigodes — CHANGES_REQUESTED** [review]
Great job overall 👏 . I left some items that require attention, including snippets we should consider porting over.

File: docs/api/doc-block-story.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+Determines whether the story is rendered `inline` (in the same browser frame as the other docs content) or in an iframe.
+
+### `meta`
+
+Type: CSF file exports
+
+Specifies the CSF file to which the story is associated.
+
+You can render a story from a CSF file that you haven’t attached to the MDX file (via `Meta`) by using the `meta` prop. Pass the **full set of exports** from the CSF file (not the default export!).
```
[the remark this answers is not in the document]
Comment (kylegach) [reply docs/api/doc-block-story.md:98]: My thought here was that it's better to start the sentence with the "why" rather than the "how". Especially since the "how" is already implied by this paragraph being under the `meta` heading.

But if you feel strongly, I'm happy to update it.

File: docs/api/doc-block-unstyled.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
+---
+title: 'Unstyled'
+---
+
+The `Unstyled` block is a special block that disables Storybook's default styling in MDX docs wherever it is added.
+
+By default, most elements (like `h1`, `p`, etc.) in docs have a few default styles applied to ensure the docs look good. However, sometimes you might want some of your content to not have these styles applied. In those cases, wrap the content with the `Unstyled` block to remove the default styles.
```
[the remark this answers is not in the document]
Comment (kylegach) [reply docs/api/doc-block-unstyled.md:7]: I disagree. The existing phrasing is more grammatically correct, imo.

File: docs/api/doc-block-controls.md [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+Specifies which controls to exclude from table. Any controls whose name matches the regex or is part of the array will be left out.
+
+### `include`
+
+Type: `string[] | RegExp`
+
+Default: `parameters.docs.controls.include`
+
+Specifies which controls to include in the table. Any controls whose name doesn’t match the regex or is not part of the array will be left out.
+
```
[the remark this answers is not in the document]
Comment (kylegach) [reply docs/api/doc-block-controls.md:78]: I applied this construction to all of the `includes`, `excludes` option descriptions.

File: docs/writing-docs/doc-blocks.md [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+```md
+<Controls exclude={['style']}>
+```
+<!-- prettier-ignore-end -->
+
+## Available blocks
+
+Each block has a dedicated API reference page, detailing usage, available options, and technical details.
+
+If a doc block accepts parameters, it is marked with **(P)**.
```
Comment (kylegach) [inline docs/writing-docs/doc-blocks.md:130]: Really not sure about this idea. Very welcome to suggestions.

File: docs/writing-docs/doc-blocks.md [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+export const autoDocsTemplate = () => (
+  <>
+    <Title />
+    <Subtitle />
+    <Description />
+    <Primary />
+    <Controls />
+    <Stories />
+  </>
+);
```
Comment (JReinhold) [inline docs/writing-docs/doc-blocks.md:68]: unfortunately this default template is being changed - and made more complex - in https://github.com/storybookjs/storybook/pull/21178
See https://github.com/storybookjs/storybook/pull/21178#issuecomment-1439164344

I'm not sure what to do here, it's a shame that it is now not as straight forward and contains actually logic. we could just update this to match, or reference the file directly? Up to you.

File: docs/writing-docs/doc-blocks.md (hunk 13 above)
Comment (JReinhold) [reply docs/writing-docs/doc-blocks.md:130]: I'm not sure it's needed? I would think it's self-explanatory that some of them reacts to `parameters.docs.X`, some don't.

Eg. `ColorPalette` doesn't because it doesn't reference any stories/components, so there's no where to pull parameters from. I don't think a user would try that.

But then I also see that `Description` don't have a (P), which definitely reads from parameters, so maybe I'm misunderstanding the idea here.

**Review by JReinhold — APPROVED** [review]

File: docs/writing-docs/doc-blocks.md (hunk 13 above)
Comment (kylegach) [reply docs/writing-docs/doc-blocks.md:130]: `Description` is an odd duck. The content is provided via parameters, but the block isn't _configured_ that way. 🤷‍♂️ 

The thought behind marking the blocks was an attempt at helping people understand (_before_ clicking through to the API reference) which blocks can be configured. But I have another idea...

File: docs/api/doc-block-story.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (jonniebigodes) [reply docs/api/doc-block-story.md:101]: Let's leave it as is, for now. We can improve it later on.

File: docs/api/doc-block-unstyled.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jonniebigodes) [reply docs/api/doc-block-unstyled.md:7]: Let's leave it as is for now. We can always improve it later on.

**Review by jonniebigodes — APPROVED** [review]

**Comment by kylegach** [comment]
This is a docs-only change, so I'm merging despite CI failures.
